design/one-pager-class-selection.md

```diff
@@ -0,0 +1,371 @@
+dynamically provision resources:
+
+```yaml
+apiVersion: cache.gcp.crossplane.io/v1alpha2
+kind: CloudMemorystoreInstanceClass
+metadata:
+  name: gcp-dev
+  labels:
+    engineVersion: "3.2"
+    region: us-west
+specTemplate:
+  tier: STANDARD_HA
+  region: us-west2
+  memorySizeGb: 1
+  redisVersion: REDIS_3_2
+---
+apiVersion: cache.azure.crossplane.io/v1alpha2
+kind: RedisClass
+metadata:
+  name: azure-dev
+  labels:
+    engineVersion: "3.2"
+    region: us-west
+specTemplate:
+  resourceGroupName: group-westus-1
+  location: West US
+  sku:
+    name: Basic
+    family: C
+    capacity: 0
+```
+
+## Technical Implementation
+
+Under the hood, resource class selection for any resource class that did not
+specify a `classRef` or `resourceRef` would be implemented as a race between a
+series of "class scheduler" controllers.
+
+In Crossplane there is not one controller for each resource claim kind, but
+rather one controller for each possible (resource claim, managed resource)
+tuple. This allows an infrastructure stack that adds support for a managed
+resource to implement the dynamic provisioning and claim binding logic for said
+managed resource without having to touch Crossplane core. Put otherwise, the GCP
+stack can enable `CloudSQLInstance` resources to satisfy `MySQLInstance` claims
+without teaching `crossplaneio/crossplane` how to dynamically provision a
+`CloudSQLInstance`. This same pattern would be apply to class scheduler
+controllers - one controller would be one responsible for each possible
+(resource claim, resource class) tuple.
+
+In more detail:
+
+* All resource claim controllers are updated to only reconcile resources with a
+  `resourceRef` or a `classRef`.
+* A class scheduler controller is introduced for each (resource claim, resource
+  class) tuple. Its job is to set the `classRef` for resource classes that omit
+  it (and omit `resourceRef`), by using their `classSelector`.
+
+Upon the creation of a `PostgreSQLInstance` without a `classRef` or `resourceRef`:
+
+1. Every class scheduler watching for `PostgreSQLInstance` claims has a
+   reconcile queued. This example discusses the `(PostgreSQLInstance,
+   CloudSQLInstanceClass)` scheduler , but `(PostgreSQLInstance,
+   RDSInstanceClass)` and `(PostgreSQLInstance, MysqlServerClass)` schedulers
+   would run through the process in parallel.
+1. The scheduler lists all `CloudSQLInstanceClass` resources that matched the
+   `PostgresSQLInstance` claim's label selectors.
+1. If no `CloudSQLInstanceClass` matched the labels, the reconcile is done.
+   Otherwise, one of the matching `CloudSQLInstanceClass` resources is selected
+   at random.
+1. The scheduler sets the `classRef` of the `PostgreSQLInstance` to the selected
+   `CloudSQLInstanceClass`. If two controllers try to set the `classRef` at the
+   same time one will fail to commit the change due to the `PostgreSQLInstance`
+   claim's resource version having changed since it was read.
```

Comment:
I initially thought we might need to artificially jitter this process to avoid conflicts, but @suskin pointed out that's not necessary. In retrospect, it may be worth jittering the process simply to increase randomness. It's likely that in cases where one scheduler controller needs to list and randomly select from 100 resource classes while another only has to list and select from 1 resource class the controller doing the least work would consistently win the race.

Reply:
I feel like it's not random if it depends on how much it takes to execute a block of code. But, given the controller-tuples structure, I see that it may not be technically feasible to build a shared list of classes and choose from that list. On the other hand, fallback to random doesn't seem to be a very frequent case to me, but, I think it's worth opening an issue to track a fix for that bias.